spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/DistributionAndOrderingUtils.scala

```diff
@@ -0,0 +1,162 @@
+import org.apache.spark.sql.catalyst.expressions.IcebergBucketTransform
+import org.apache.spark.sql.catalyst.expressions.IcebergDayTransform
+import org.apache.spark.sql.catalyst.expressions.IcebergHourTransform
+import org.apache.spark.sql.catalyst.expressions.IcebergMonthTransform
+import org.apache.spark.sql.catalyst.expressions.IcebergYearTransform
+import org.apache.spark.sql.catalyst.expressions.NamedExpression
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.plans.logical.RepartitionByExpression
+import org.apache.spark.sql.catalyst.plans.logical.Sort
+import org.apache.spark.sql.connector.catalog.CatalogV2Implicits
+import org.apache.spark.sql.connector.expressions.BucketTransform
+import org.apache.spark.sql.connector.expressions.DaysTransform
+import org.apache.spark.sql.connector.expressions.Expression
+import org.apache.spark.sql.connector.expressions.FieldReference
+import org.apache.spark.sql.connector.expressions.HoursTransform
+import org.apache.spark.sql.connector.expressions.IdentityTransform
+import org.apache.spark.sql.connector.expressions.MonthsTransform
+import org.apache.spark.sql.connector.expressions.NamedReference
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.connector.expressions.YearsTransform
+import org.apache.spark.sql.connector.iceberg.distributions.ClusteredDistribution
+import org.apache.spark.sql.connector.iceberg.distributions.Distribution
+import org.apache.spark.sql.connector.iceberg.distributions.OrderedDistribution
+import org.apache.spark.sql.connector.iceberg.distributions.UnspecifiedDistribution
+import org.apache.spark.sql.connector.iceberg.expressions.NullOrdering
+import org.apache.spark.sql.connector.iceberg.expressions.SortDirection
+import org.apache.spark.sql.connector.iceberg.expressions.SortOrder
+import org.apache.spark.sql.internal.SQLConf
+
+object DistributionAndOrderingUtils {
+
+  def prepareQuery(
+      requiredDistribution: Distribution,
+      requiredOrdering: Seq[SortOrder],
+      query: LogicalPlan,
+      conf: SQLConf): LogicalPlan = {
+
+    val resolver = conf.resolver
+
+    val distribution = requiredDistribution match {
+      case d: OrderedDistribution =>
+        d.ordering.map(e => toCatalyst(e, query, resolver))
+      case d: ClusteredDistribution =>
+        d.clustering.map(e => toCatalyst(e, query, resolver))
+      case _: UnspecifiedDistribution =>
+        Array.empty[catalyst.expressions.Expression]
+    }
+
+    val queryWithDistribution = if (distribution.nonEmpty) {
+      val numShufflePartitions = conf.numShufflePartitions
+      // the conversion to catalyst expressions above produces SortOrder expressions
+      // for OrderedDistribution and generic expressions for ClusteredDistribution
+      // this allows RepartitionByExpression to pick either range or hash partitioning
+      RepartitionByExpression(distribution, query, numShufflePartitions)
```

Comment:
When `WRITE_DISTRIBUTION_MODE = range`, before this PR, Logical Plan is

```
Sort [dt#8 ASC NULLS FIRST, v#7 ASC NULLS FIRST], true
+- Repartition 2000, true
   +- MergeInto 
```
After this PR, Logical Plan is

```
Sort [dt#8 ASC NULLS FIRST, v#7 ASC NULLS FIRST], false
+- RepartitionByExpression [dt#8], 2000
   +- Repartition 2000, true
      +- MergeInto
```

In my opinion, the conversion of `global sort` to `local sort + range partitioning` is correct, but here we need to consider the `CollapseRepartition` rule in Spark `Optimizer`. In this case, this rule will eliminate the `Repartition 2000, true` node.

Please take a look here, thanks @aokolnychyi @rdblue